gpbft/gpbft.go

```diff
@@ -351,6 +376,44 @@ func (i *instance) receiveOne(msg *GMessage) error {
 	return i.tryCurrentPhase()
 }
 
+// shouldSkipToRound determines whether to skip to round, and justification
+// either for a value to sway to, or of COMMIT bottom to justify our own
+// proposal. Otherwise, it returns nil justification and false. When a
+// justification exists for a value to sway to this function does sway to that
+// value. The caller is then expected to do skip to the given round.
+//
+// See: skipToRound.
+func (i *instance) shouldSkipToRound(round uint64, state *roundState) (*Justification, bool) {
+
+	// Check if the given round is ahead of current round and this instance is not in
+	// QUALITY nor DECIDE phase as dedicated by FIP-0086.
+	//
+	// Note that sipping ahead from QUALITY phase does not violate the correctness
+	// proof of gPBFT. The fact that QUALITY always terminates for a participant
+	// means this participant will eventually jump, but it may do it earlier.
+	// However, if a node skips to future rounds without executing the QUALITY phase
+	// it will only help reaching consensus for bottom.
+	//
+	// Future work may consider skipping ahead from QUALITY for faster census in
+	// certain scenarios. For now, this implementation conforms to the FIP.
+	if round <= i.round || i.phase == QUALITY_PHASE || i.phase == DECIDE_PHASE {
+		return nil, false
+	}
+	proposal := state.converged.FindMaxTicketProposal(i.powerTable)
+	if proposal.Justification == nil {
+		return nil, false
+	}
+	if !state.prepared.ReceivedFromWeakQuorum() {
+		return nil, false
+	}
+	if proposal.Justification.Vote.Step == PREPARE_PHASE {
+		i.log("⚠️ swaying from %s to %s by skip to round %d", &i.proposal, proposal.Chain, i.round)
+		i.candidates = append(i.candidates, proposal.Chain)
+		i.proposal = proposal.Chain
+	}
```

Comment:
I would probably do this in skipToRound so that this function, which sounds like a pure predicate, doesn't have side effects. This is pretty minor in the context though.

Reply:
Ah I thought [previous suggestion](https://github.com/filecoin-project/go-f3/pull/276#discussion_r1612488201) was to move it to `shouldSkipToRound`. That's where I had it initially 👍